e2e/nomostest/sync-status-watch-controller.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nomostest
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/pkg/kinds"
+
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/core/k8sobjects"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// StatusWatchNamespace is the namespace for sync status watch resources.
+	StatusWatchNamespace = "sync-status-watch"
+	// StatusWatchName is the name for sync status watch resources.
+	StatusWatchName = "sync-status-watch"
+)
+
+const sswImage = testing.TestInfraArtifactRepositoryAddress + "/sync-status-watch-controller:v1.0.0-539a1ddd"
+
+// SetupSyncStatusWatchController builds and deploys the sync status watch controller.
+func SetupSyncStatusWatchController(nt *NT) error {
+	nt.T.Logf("applying sync status watch controller manifest")
+	if err := nt.KubeClient.Create(testSyncStatusWatchNamespace()); err != nil {
+		return err
+	}
+	if err := nt.KubeClient.Create(testSyncStatusWatchServiceAccount()); err != nil {
+		return err
+	}
+
+	if err := installSyncStatusWatchController(nt); err != nil {
+		nt.describeNotRunningTestPods(StatusWatchNamespace)
+		return fmt.Errorf("waiting for sync status watch controller to become available: %w", err)
+	}
+	return nil
+}
+
+// TeardownSyncStatusWatchController removes the sync status watch controller.
+func TeardownSyncStatusWatchController(nt *NT) error {
+	nt.T.Log("tearing down sync status watch controller")
+	objs := testSyncStatusWatchObjects()
+	for _, o := range objs {
+		if err := nt.KubeClient.Delete(o); err != nil && !apierrors.IsNotFound(err) {
+			nt.T.Logf("Error deleting %T %s/%s: %v", o, o.GetNamespace(), o.GetName(), err)
+		}
+	}
+	if err := nt.KubeClient.Delete(testSyncStatusWatchNamespace()); err != nil && !apierrors.IsNotFound(err) {
+		nt.T.Logf("Error deleting Namespace: %v", err)
+	}
+
+	nt.T.Log("Sync status watch controller teardown complete")
+	return nil
+}
+
+func installSyncStatusWatchController(nt *NT) error {
+	objs := testSyncStatusWatchObjects()
```

Comment:
I'd suggest using the real manifests defined in the example directory instead of defining these k8s objects in a duplicate place here. That would give greater confidence as we would be testing with the real manifests, and ensure there hasn't been any drift between the real manifests and the test manifests

Reply:
Changed the setup and teardown to: get the manifest, replace the image placeholder with test image, create a temp file so that the temp file can be applied / delete using kubectl. Thought about putting the test image directly into the manifest but it would disturb the yaml when serving as part of user guide, so chose the temp file as an intermediate solution.